tutorial4.md

```diff
@@ -0,0 +1,424 @@
+-k8s-tutorial_kube-system_0e03e38505f7290852ed04a5db4b9d73_0
+72d77b649ca3        k8s.gcr.io/pause:3.1   "/pause"                 About a minute ago   Up About a minute                       k8s_POD_kube-apiserver-k8s-tutorial_kube-system_0e03e38505f7290852ed04a5db4b9d73_0
+```
+
+## Connecting to the API Server
+
+The standard way of connecting to a Kubernetes cluster is a YAML file called a `kubeconfig`.
+This file contains any number of credentials (which can be passwords, client certificates, or cloud provider tokens),
+as well as an address to connect to.
+
+In our case, it'll be a local IP addresses and credentials signed by by the apiserver certificate authority.
+
+We can get kubeadm to generate this:
+
+```console
+$ sudo kubeadm init phase kubeconfig admin
+[kubeconfig] Writing "admin.conf" kubeconfig file
+```
+
+The file is generated as `/etc/kubernetes/admin.conf`, not readable by non-root users.
+Let's move them somewhere more convenient:
+
+```shell
+mkdir ~/.kube
+sudo cp /etc/kubernetes/admin.conf ~/.kube/config
+sudo chown $(whoami) ~/.kube/config
+```
+
+### `kubectl`
+
+If you haven't already, you should install [`kubectl`][kubectl].
+`kubectl` is utility used to access, configure, and administer kubernetes clusters.
+
+[kubectl]: https://kubernetes.io/docs/reference/kubectl/overview/
+
+You should be able to install it using `apt`:
+
+```shell
+sudo apt install kubectl
+```
+
+And now you should be able to query the api server:
+
+```console
+$ kubectl get cm --all-namespaces
+NAMESPACE     NAME                                 DATA   AGE
+kube-system   extension-apiserver-authentication   1      16m
+```
+
+## Reconnect kubelet
+
+Right now, the `kubelet` isn't connected to the apiserver, so we won't be able to run any pods.
+Let's fix that.
+
+### Bootstrap credentials
+
+During [Tutorial 1][tut1], we commented out the first line of the `kubeadm`-provided `kubelet`.
+That line pertained to [bootstrap credentials][bootstrap], and before we can uncomment it we need to generate those.
+
+[tut1]: tutorial1.md#your-first-config-file
+[bootstrap]: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/
+
+`kubeadm` can create a bootstrap kubeconfig:
+
+```console
+$ sudo kubeadm init phase kubeconfig kubelet
+[kubeconfig] Writing "kubelet.conf" kubeconfig file
+```
+
+Now we can uncomment that file:
+
+```shell
+sudoedit /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+sudo systemctl daemon-reload
+sudo systemctl restart kubelet
+```
+
+If all's well, you should now be able to see the pods the kubelet is running for you:
+
+```shell
+$ kubectl get pods --namespace kube-system
+NAME                          READY   STATUS    RESTARTS   AGE
+etcd-k8s-tutorial             1/1     Running   0          2m10s
+kube-apiserver-k8s-tutorial   1/1     Running   0          2m10s
+```
```

Comment:
"if you're curious, you can look at the contents of `/etc/kubernetes/kubelet.conf` now that it's created (it's not necessary for `/etc/kubernetes/bootsrap-kubelet.conf` to exist as long as kubelet.conf exists)..."